src/lib.rs

```diff
@@ -511,3 +556,315 @@ where
+
+        // Since we have synced up with the currently assigned partitiions and offsets, we can mark
+        // ourselves clean.
+        self.dirty = false;
+
+        Ok(())
+    }
+
+    pub fn is_tracking(&self, partition: DataTypePartition) -> bool {
+        self.partitions.contains(&partition)
+    }
+
+    pub fn mark_dirty(&mut self) {
+        self.dirty = true;
+    }
+}
+
+pub struct ConsumedBuffers {
+    pub values: Vec<Value>,
+    pub partition_offsets: HashMap<DataTypePartition, DataTypeOffset>,
+}
+
+pub struct ValueBuffer {
+    last_offset: Option<DataTypeOffset>,
+    values: Vec<Value>,
+}
+
+impl ValueBuffer {
+    fn new() -> Self {
+        Self {
+            last_offset: None,
+            values: Vec::new(),
+        }
+    }
+
+    fn add(&mut self, value: Value, offset: DataTypeOffset) {
+        self.last_offset = Some(offset);
+        self.values.push(value);
+    }
+
+    fn reset(&mut self, offset: Option<DataTypeOffset>) {
+        self.last_offset = offset;
+        self.values.clear();
+    }
+
+    fn consume(&mut self) -> (Vec<Value>, DataTypeOffset) {
+        let consumed = (self.values.drain(0..).collect(), self.last_offset.unwrap());
+        self.last_offset = None;
+        consumed
+    }
+
+    fn non_empty(&self) -> bool {
+        self.last_offset != None
+    }
+}
+
+pub struct Context {
+    partition_assignment: Arc<Mutex<PartitionAssignment>>,
+}
+
+impl ClientContext for Context {}
+
+impl ConsumerContext for Context {
+    fn pre_rebalance(&self, rebalance: &Rebalance) {
+        info!("Pre rebalance {:?}", rebalance);
+    }
+
+    fn post_rebalance(&self, rebalance: &Rebalance) {
+        info!("Post rebalance {:?}", rebalance);
+
+        match rebalance {
+            Rebalance::Assign(tpl) => {
+                info!("Received new partition assignment list");
+                let partitions = partition_vec_from_topic_partition_list(tpl);
+                let partitions = Arc::new(partitions);
+                on_rebalance_assign(self.partition_assignment.clone(), partitions.clone());
+            }
+            Rebalance::Revoke => {
+                info!("Partition assignments revoked");
+                on_rebalance_revoke(self.partition_assignment.clone());
+            }
+            Rebalance::Error(e) => {
+                warn!(
+                    "Unexpected Kafka error in post_rebalance invocation {:?}",
+                    e
+                );
+            }
+        }
+    }
+}
+
+fn partition_vec_from_topic_partition_list(
+    topic_partition_list: &TopicPartitionList,
+) -> Vec<DataTypePartition> {
+    topic_partition_list
+        .to_topic_map()
+        .iter()
+        .map(|((_, p), _)| *p)
+        .collect()
+}
+
+fn on_rebalance_assign(
+    partition_assignment: Arc<Mutex<PartitionAssignment>>,
+    partitions: Arc<Vec<DataTypePartition>>,
+) {
+    let _fut = tokio::spawn(async move {
```

Comment:
`PartitionAssignment` is wrapped in `Arc<tokio::sync::Mutex<>>`.  I had to use the `tokio::sync::Mutex` because the `std::sync::Mutex` returns a `MutexGuard` that cannot wrap async calls. I have to hold the lock open while making async WAL calls though. The tokio lock requires an `await`, but the `rdkafka` `ConsumerContext` trait rebalance methods are non-async, so I had to bridge the gap.

Reply:
ha, ok, that makes sense. thanks for the explanation.